src/jit/flowgraph.cpp

```diff
@@ -18969,6 +18969,18 @@ void Compiler::fgSetBlockOrder()
             // JIT64 does.
             genInterruptible = true;
         }
+
+#if defined(_TARGET_ARM_) || defined(_TARGET_ARM64_)
```

Comment:
I think this is the wrong location to check for this, although it probably works in most cases.

For the "hasTailCalls" bool, you only care if a function has any tailcalls at all, after it has been completely generated. If possible, this means we should delay the checking for tailcall existence until as late as possible, to prevent the case where we initially think we'll generate a tailcall, then later we change our mind due to some kind of optimization.

Also, we should clearly define what "hasTailCalls" means. E.g., the "normal" prolog/epilog sequence (for arm32) is "push {...,lr} ... pop {..., pc}", but we can also generate "push {..., lr} ... pop {..., lr}; bx lr". Are both variations ok?

And what about the JMP IL instruction? That is very similar to tailcall, in that we branch to the target function address, and (I think) LR remains live as the return address of the called target function.

I think you might want to put code in the `_TARGET_ARMARCH_` version of `genFnEpilog()` (about line 9574 of codegencommon.cpp), in particular, within the `if (jmpEpilog)` clause:
```
if (jmpEpilog)
{
    hasTailCalls = true;
    ...
```
Note that this code is run for both JMP and fast tail call.

Reply:
Thanks @BruceForstall for the detailed review. When unhijack happens when the target thread is suspended at the `bx lr` (which is the only way to return on arm64, since there is no `pop pc`) and the function that was hijacked is the current function, it just restores the original return address to a location on the stack that's at an address that's less than the SP. It has no effect and causes no harm. When the target thread is resumed, it returns to the OnHijackTripThread. It pushes LR to stack and then calls the OnHijackWorker. This function places the original return address to the place where the OnHijackTripThread stored the LR and marks the thread as not hijacked. Then it does its hijacking related job and finally returns to the OnHijackTripThread  that in turn returns to the original address that was hijacked. So it fixes the fact that the unhijack modified a wrong location.
The tail call case is different. We end up entering the tail call target with LR set to the OnHijackTripThread. Now there are several cases that can happen w.r.t. where the unhijack happens:
1. The target thread is executing the tail call target and it hasn't allocated the stack frame yet. The unhijack has no effect, since it restores the original return address to a location where it will either get overwritten by the current function's frame or to a location that's at lower address than the lowest address of the current function's frame.
2. The target thread is executing the tail call target and it has already allocated the stack frame. This has three sub-cases:
  a. The tail call caller was storing the LR at a location that's inside of the current function's frame at an address different from the location where the tail call target stores the LR. In this case, unhijack corrupts the stack frame.
  b. The tail call caller was storing the LR at a location that's at lower address than the lowest address of the current function's frame. In this case, the unhijack has no effect.
  c. The tail call caller was storing the LR at a location that's at the same address address as the address where the LR is stored in the current function's frame. In this case, the unhijack actually works correctly.
3. The target thread is executing a function that was transitively called by the tail call target. Then we are in a situation similar to (2), but it would also cause corruption if the original location of the hijacked return address was inside one of the frames of the call chain from the tail call target and the current function. It could even cause more subtle effect in case that location by a chance matched a location of LR stored by another function in the call chain. Then the unhijack would cause the current function to return to original return address of the tail call caller instead of the real caller of the current function.

Now all of the cases described above where I've said that the unhijack would have no effect would actually "heal" the issue by returning from the tail call target to the OnHijackTripThread that would then return to the original return address that was saved.
But only if another hijack didn't happen before it returned from the tail call target. If another hijack happened, then the original return address would be lost, overwritten by the new one. If the hijack happened in the tail call target, the thread would end up running the OnHijackTripThread in an infinite loop, since it would store the OnHijackTripThread as the original return address. If the hijack happened in some function transitively called by the tail call target, then the tail call target would still return to the OnHijackTripThread, but that one would then return to the caller of the function that was hijacked, likely crashing, hanging or doing other interesting stuff.

Reply:
Thanks for the detailed explanation. I guess this never happened on ARM because the JIT hasn't implemented fast tail call optimization (although #13828 has been waiting, to enable it). On ARM64, I guess we just haven't noticed it (or tested it well). Looks like test JIT/Methodical/tailcall_v4/hijacking/hijacking.sh is disabled for Linux in testsFailingOnArm64.txt and for Windows in tests\arm64\tests.lst, despite #4122 claiming the original failure was fixed.

I'm guessing your case 2c is the common case, which would greatly reduce the incidence of this bug. I presume cases of multiple hijacks, where we would "lose" the original address, are also rare, since once we hijack we wait for a while for it to kick in, and presumably most code returns via the hijacked address in that time.

As a matter of principle, at least, it seems like the unhijack code should not be writing to the stack (to restore the stack-based LR to its original value) at any address below the current SP. Presumably the VM has the context, and can compare the hijack address (m_ppvHJRetAddrPtr) against SP. It could notice that the hijack won't work in this case (such as the "pop {lr}; <unhijack>; bx lr" case).

The hijack code saves the `MethodDesc*` of the function when the hijack is installed. Presumably that could be used in the unhijack case. E.g., assume the hijack happened in a tailcall caller, and now we are in a nested call tree within the tailcall target. If we go to unhijack, we could walk the stack looking for a function that contains the hijacked return address (m_pvHJRetAddr). If we reach this address, and the frame one lower on the stack (call it the "unhijack frame") isn't the saved `MethodDesc*`, then we have a mismatch. We could use the stack walker to tell us where in the "unhijack frame" LR is stored, and unhijack to that location instead of the saved location. So, for `A --call--> B --tailcall--> C --call--> D --call--> E`, with hijack in B, and unhijack in E: the stack frame looks like: ` A C D E ` (since the stack frame for B disappears). If we walk the stack, when we get to A, the saved return address will be in the code for function A. But we never found the saved `MethodDesc*` corresponding to B. So, we know B is gone. We could figure out where the return address is stored in the stack frame for C, and set it to the return address in A. Presumably, the saved LR in the stack frame for C must be OnHijackTripThread. (Actually, can we walk the stack if the saved LR is "incorrect"?)

Perhaps this is too complex (or there are holes in my logic).